# DS Automation Assignment

# Importing Modules

In [1]:
from tpot import TPOTClassifier
import pandas as pd

from sklearn.model_selection import train_test_split

import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline, make_union
from sklearn.preprocessing import PolynomialFeatures
from tpot.builtins import StackingEstimator
from tpot.export_utils import set_param_recursive

C:\Users\Sharadha\AppData\Local\Programs\Python\Python310\lib\site-packages\tpot\builtins\__init__.py:36: UserWarning: Warning: optional dependency `torch` is not available. - skipping import of NN models.
  warnings.warn("Warning: optional dependency `torch` is not available. - skipping import of NN models.")


# Loading data and Prepping it

In [2]:
df = pd.read_csv('prepped_churn.csv')
df.head()

Unnamed: 0  tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
0           0       1             1         0              3           29.85   
1           1      34             0         1              1           56.95   
2           2       2             0         0              1           53.85   
3           3      45             1         1              4           42.30   
4           4       2             0         0              3           70.70   

   TotalCharges  Churn  TotalCharges_tenure_ratio  
0         29.85      1                  29.850000  
1       1889.50      1                  55.573529  
2        108.15      0                  54.075000  
3       1840.75      1                  40.905556  
4        151.65      0                  75.825000

In [3]:
del df['Unnamed: 0']
df.head()

tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
0       1             1         0              3           29.85   
1      34             0         1              1           56.95   
2       2             0         0              1           53.85   
3      45             1         1              4           42.30   
4       2             0         0              3           70.70   

   TotalCharges  Churn  TotalCharges_tenure_ratio  
0         29.85      1                  29.850000  
1       1889.50      1                  55.573529  
2        108.15      0                  54.075000  
3       1840.75      1                  40.905556  
4        151.65      0                  75.825000

In [4]:
features = df.drop('Churn', axis=1)
targets = df['Churn']

x_train, x_test, y_train, y_test = train_test_split(features, targets, stratify=targets, random_state=42)

# Tpot Classifier

We create an instance of tpot classifier with some custom parameters. We give generations as 3 to have 3 iterations to find the best pipeline. We can then use this instance and fit our training data. The classifier then gives us the best accuracy score for each iteration performed. In the end, it shows us the pipeline that best fit our data.

In [5]:
#create classifier instance
tpot = TPOTClassifier(generations=3, population_size=20, cv=5,random_state=42, verbosity=2)

In [6]:
#fit training data into tpot classifier
tpot.fit(x_train, y_train)

Optimization Progress:   0%|          | 0/80 [00:00<?, ?pipeline/s]


Generation 1 - Current best internal CV score: 0.7956043778159482

Generation 2 - Current best internal CV score: 0.7956043778159482

Generation 3 - Current best internal CV score: 0.7961727384731602

Best pipeline: ExtraTreesClassifier(input_matrix, bootstrap=False, criterion=gini, max_features=0.35000000000000003, min_samples_leaf=18, min_samples_split=9, n_estimators=100)


TPOTClassifier(generations=3, population_size=20, random_state=42, verbosity=2)

We got the pipeline that best fit our data as ExtraTreesClassifier, with an accuracy score of 79.6%.
Next, we check how this model performs with our test data and get the score.

In [7]:
#obtaining performance score on test data
tpot.score(x_test,y_test)

C:\Users\Sharadha\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_scorer.py:765: FutureWarning: sklearn.metrics.SCORERS is deprecated and will be removed in v1.3. Please use sklearn.metrics.get_scorer_names to get a list of available scorers and sklearn.metrics.get_metric to get scorer.
  warnings.warn(


0.7935153583617748

Lastly, we export the pipeline we created as a .py file. We are calling this file as 'tpot_exported.py'.

In [8]:
#export pipeline
tpot.export('tpot_exported.py')

# Summary

In this file, we use tpot classifier to obtain the best pipeline and export this pipeline. 
We first loaded in our prepped churn data. We then initialised an instance of TPOTClassifier with a few custom parameters. We fit the training data in our classifier and after the 3 iterations, obtained the best pipeline as an ExtraTreesClassifier with some custom parameters. We got the accuracy as 79.6%. We tested the model with our test data and got a score of 79.3%. As there is only a small difference between the accuracy obtained for the train and test data, we can clarify that the data was not overfit. We then exported this model into our system. 